File: cses/graphs/shortest_routes2/sr2.cpp
#include <bits/stdc++.h>

using namespace std;

#define endl '\n'

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  double n, m, q;
  cin >> n >> m >> q;
  double inf = numeric_limits<double>::infinity();
  vector<vector<double>> dist(n + 1, vector<double>(n + 1, inf));

  for (int i = 0; i < m; i++) {
    double fr, to, w;
    cin >> fr >> to >> w;
    dist[fr][to] = min(dist[fr][to], w);
    dist[to][fr] = min(dist[to][fr], w);
  }
  for (int i = 1; i <= n; i++) {
    dist[i][i] = 0;
  }

  for (int via = 1; via <= n; via++) {
    for (int fr = 1; fr <= n; fr++) {
      for (int to = 1; to <= n; to++) {
        dist[fr][to] = min(dist[fr][to], dist[fr][via] + dist[via][to]);
      }
    }
  }

  for (int i = 0; i < q; i++) {
    double fr, to;
    cin >> fr >> to;
    double D = dist[fr][to];
    D = D == inf ? -1 : D;
    cout << (long long)D << endl;
  }

  return 0;
}

File: cses/graphs/shortest_routes2/sr2.py
from itertools import permutations
import sys
from pprint import pprint
from collections import deque, defaultdict
from functools import lru_cache

sys.setrecursionlimit(1 << 24)

def ints():
    return map(int, input().split())


n, m, q = ints()

dist = [[float("inf")] * n for _ in range(n)]

for i in range(m):
    fr, to, w = ints()
    fr -= 1
    to -= 1
    dist[fr][to] = min(dist[fr][to], w)
    dist[to][fr] = min(dist[to][fr], w)

for via in range(n):
    for fr in range(n):
        for to in range(n):
            dist[fr][to] = min(dist[fr][to], dist[fr][via] + dist[via][to])

for i in range(q):
    fr, to = ints()
    fr -= 1
    to -= 1
    D = dist[fr][to]
    if D == float("inf"):
        D = -1
    print(D)
